workflow/controller/artifact_gc.go

```diff
@@ -64,6 +64,12 @@ func (woc *wfOperationCtx) garbageCollectArtifacts(ctx context.Context) error {
 		}
 	}
 
+	if artifactsWriting {
+		// If we don't do this, a race condition can occur where only some artifacts are garbage collected and finalizers are prematurely removed.
+		woc.log.Debug("Skipping garbage collection completion due to in-progress artifact reporting.")
+		return nil
+	}
+
```

Comment:
instead of passing in `artifactsWriting`, can we remove `processArtifactGCCompletion()` from this function, and have the caller conditionally call `processArtifactGCCompletion()`?

Reply:
actually, wait...now I'm thinking about whether we need to be preventing `processArtifactGCStrategy()` from running as well, since it sort of relies on the Artifacts being in the Node Outputs as well.

Reply:
I'm thinking the part we should do early on is the adding of the finalizer (section starting on line 42). Everything after that we shouldn't. Can we break it into two methods? Maybe the first can just be called `addArtifactGCFinalizer()` or something and the second one can continue to remain as `garbageCollectArtifacts()`? something like that?

Reply:
Sounds good. I'll do that.